pkg/kubelet/kubelet.go

```diff
@@ -1731,6 +1719,155 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
 	return nil
 }
 
+// syncTerminatingPod is expected to terminate all running containers in a pod. Once this method
+// returns without error, the pod can be safely cleaned up. If runningPod is passed, we perform
+// no status updates.
+func (kl *Kubelet) syncTerminatingPod(pod *v1.Pod, podStatus *kubecontainer.PodStatus, runningPod *kubecontainer.Pod, gracePeriod *int64, podStatusFn func(*v1.PodStatus)) error {
+	klog.InfoS("DEBUG: syncTerminatingPod enter", "pod", klog.KObj(pod), "podUID", pod.UID)
+	defer klog.InfoS("DEBUG: syncTerminatingPod exit", "pod", klog.KObj(pod), "podUID", pod.UID)
+
+	// when we receive a runtime only pod (runningPod != nil) we don't need to update the status
+	// manager or refresh the status of the cache, because a successful killPod will ensure we do
+	// not get invoked again
+	if runningPod != nil {
+		// we kill the pod with the specified grace period since this is a termination
+		if gracePeriod != nil {
+			klog.V(4).InfoS("Pod terminating with grace period", "pod", klog.KObj(pod), "podUID", pod.UID, "gracePeriod", *gracePeriod)
+		} else {
+			klog.V(4).InfoS("Pod terminating with grace period", "pod", klog.KObj(pod), "podUID", pod.UID, "gracePeriod", nil)
+		}
+		if err := kl.killPod(pod, *runningPod, gracePeriod); err != nil {
+			kl.recorder.Eventf(pod, v1.EventTypeWarning, events.FailedToKillPod, "error killing pod: %v", err)
+			// there was an error killing the pod, so we return that error directly
+			utilruntime.HandleError(err)
+			return err
+		}
+		klog.V(4).InfoS("Pod termination stopped all running orphan containers", "pod", klog.KObj(pod), "podUID", pod.UID)
+		return nil
+	}
+
+	// In order to safely terminate a pod, we need to know the set of sandboxes that the last
+	// syncPod method created as well as running containers - but we also wish to do so with
+	// the lowest latency possible, which means we don't want the pod worker to have to wait
+	// to call syncTerminatingPod until container status is recalculated. Instead, the first
+	// invocation of syncTerminatingPod on a given pod may send a nil podStatus to allow us
+	// to more quickly guarantee shutdown of everything the last syncPod created.
+	if podStatus == nil {
```

Comment:
You're right.  That update also needs to be conditional on timestamp (I think) - since updateCache() is not done under lock.  Really, only PLEG should be able to invoke Set(), and PLEG needs to either rely on cache to enforce timestamp ordering, or PLEG itself needs to enforce timestamp ordering (but since cache has the lock and the data, it's probably cache for SetIfNewer()).

Reply:
See https://github.com/kubernetes/kubernetes/pull/102344/commits/416b068dab13bc7710c1cce9b5f3072bf93ae21d which i think is the correct way because only PLEG can control happens-before on status read.

Are we guaranteed to have the updated cache by the time the next syncTerminating is called?  I don't think we do

Reply:
Agree that we should not directly update pod cache here